src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java

```diff
@@ -9,20 +16,64 @@
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 import java.io.IOException;
+import java.text.NumberFormat;
+import java.text.ParseException;
 
 public class ProductDetailsPageServlet extends HttpServlet {
     private ProductDao productDao;
+    private CartService cartService;
+    private RecentlyViewedService recentlyViewedService;
+    private boolean productAddedToCart;
 
     @Override
     public void init(ServletConfig config) throws ServletException {
         super.init(config);
         productDao = ArrayListProductDao.getInstance();
+        cartService = DefaultCartService.getInstance();
+        recentlyViewedService = DefaultRecentlyViewedService.getInstance();
     }
 
     @Override
     protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
-        String productId = request.getPathInfo();
-        request.setAttribute("product", productDao.getProduct(Long.valueOf(productId.substring(1))));
+        Long productId = parseProductId(request);
+        if (request.getParameter("message") != null && productAddedToCart != false) {
```

Comment:
1) != false can be reaplced with !productAddedToCart 2) What is the purpose of this mechanism? Better to display just param value in jsp instead of such logic. Also please avoid usage of such flags in future, especially in high level classes (servlet, controller), it's very bad practice, but if you need this you can store such variable in some user specific context, in our case it's HttpSession. Think about situation from multi threading view, when there are 2 requests coming in one time, you will have productAddedToCart=true, but it can be incorrect.